lib/surface/compiler/parser.ex

```diff
@@ -102,32 +96,94 @@ defmodule Surface.Compiler.Parser do
+  defp handle_token(
+         [{:block_open, name, {:expr, expr, expr_meta}, meta} | rest],
+         buffers,
+         state
+       ) do
+    attrs = [{:root, {:expr, expr, expr_meta}, expr_meta}]
+    handle_token([{:block_open, name, attrs, meta} | rest], buffers, state)
+  end
+
+  defp handle_token([{:block_open, name, nil, meta} | rest], buffers, state) do
+    handle_token([{:block_open, name, [], meta} | rest], buffers, state)
+  end
+
+  ###
+
+  defp handle_token([{:block_open, name, attrs, meta} = token | rest], buffers, state)
+       when name in @sub_blocks do
+    {buffers, state} = close_sub_block(token, buffers, state)
+
+    # push the current sub-block token to state
+    state = push_tag(state, {:block_open, name, attrs, meta})
+
+    # create a new buffer for the current sub-block
+    buffers = [[] | buffers]
+
+    handle_token(rest, buffers, state)
+  end
+
+  defp handle_token([{:block_open, name, _attrs, _meta} = token | rest], buffers, state)
+       when name in @blocks do
+    state = push_tag(state, token)
+    # create a new buffer for the node
+    buffers = [[] | buffers]
+    handle_token(rest, buffers, state)
+  end
+
+  defp handle_token([{:block_open, name, _attrs, meta} | _], _buffers, _state) do
+    blocks = Helpers.list_to_string("block is", "blocks are", @blocks ++ @sub_blocks)
+    raise parse_error("unknown `{##{name}}` block. Available #{blocks}", meta)
+  end
+
   defp handle_token(
-         [{:tag_close, _name, _meta} = token | _] = tokens,
+         [{:block_close, _name, _meta} = token | _] = tokens,
          buffers,
-         %{tags: [{:tag_open, name, _, _} | _]} = state
+         %{tags: [{:block_open, name, _, _} | _]} = state
        )
        when name in @sub_blocks do
     {buffers, state} = close_sub_block(token, buffers, state)
     handle_token(tokens, buffers, state)
   end
 
-  defp handle_token([{:tag_close, name, _meta} | rest], buffers, state) do
-    {{:tag_open, _name, attrs, meta}, state} = pop_matching_tag(state, name)
+  defp handle_token([{:block_close, name, _meta} = token | rest], buffers, state)
+       when name in @blocks do
+    {{:block_open, _name, attrs, meta}, state} = pop_matching_tag(state, token)
 
     # pop the current buffer and use it as children for the node
     [buffer | buffers] = buffers
-    node = {name, transtate_attrs(attrs), Enum.reverse(buffer), to_meta(meta)}
+    node = {:block, name, transtate_attrs(attrs), Enum.reverse(buffer), to_meta(meta)}
     buffers = push_node_to_current_buffer(node, buffers)
     handle_token(rest, buffers, state)
   end
 
-  # IF there's a previous sub-block defined. Close it.
-  defp close_sub_block(_token, buffers, %{tags: [{:tag_open, name, attrs, meta} | tags]} = state)
+  defp handle_token([{:block_close, name, meta} | _], _buffers, _state) do
+    blocks = Helpers.list_to_string("block is", "blocks are", @blocks ++ @sub_blocks)
```

Comment:
```suggestion
    blocks = Helpers.list_to_string("block is", "blocks are", @blocks)
```
I would remove `@sub_blocks` as the user can't close a sub block with `{/sub_block_name}`, he must use the parent block name to close the whole block.

We can go further and if the user uses a subblock name to "close" a block we could give him a proper hint. For example, with the `else` sub block:
> You can only close a sub block by closing the whole block. Did you mean `{/if}` or `{/for}`?

Reply:
Good catch!